feat: move the widget creation CTA off the Entity explorer

akash-codemonk opened the pull request:
## Description

 The widget pane is moved outside the explorer. Also adds a UI to add top 5 popular widgets without having to open the pane. Most file changes includes updates to test files to update assertions with drag n drop.

### TLDR
 The widget pane is moved outside the explorer. Also adds a UI to add top 5 popular widgets without having to open the pane.

Fixes https://github.com/appsmithorg/appsmith/issues/17511


Media

https://user-images.githubusercontent.com/67054171/215044634-b18e2348-fb3a-406d-ae14-c35439dd1be8.mov




## Type of change

- New feature (non-breaking change which adds functionality)


## How Has This Been Tested?

- Manual
- Cypress

### Test Plan
> Add Testsmith test cases links that relate to this PR

### Issues raised during DP testing
https://github.com/appsmithorg/appsmith/pull/19844#issuecomment-1396604559


## Checklist:
### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag


### QA activity:
- [x] Test plan has been approved by relevant developers
- [x] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [x] Added Test Plan Approved label after reveiwing all Cypress test

akash-codemonk commented:
/ok-to-test sha=3e5fd1e

Richarex commented:
- [ ] Widget selection box doesn't goes off in 1st click on welcome tutorial page on 6th step.
i) Button icon is getting highlighted only for the 1st time when the widget tab open is it expected?
ii)The widget box lines are not aligned properly as well. Please refer to the attached video:- https://jiju8jbmwa.vmaker.com/record/OSaCRbF5Tvju6P7F

akash-codemonk commented:
/ok-to-test sha=5e59dba

akash-codemonk commented:
/ok-to-test sha=b073d85

akash-codemonk commented:
/ok-to-test sha=418881d

akash-codemonk commented:
/ok-to-test sha=c0c0197

akash-codemonk commented:
/ok-to-test sha=35c8b80

akash-codemonk commented:
/ok-to-test sha=6613205

akash-codemonk commented:
/ok-to-test sha=e889cbe

akash-codemonk commented:
/ok-to-test sha=69f2f13

akash-codemonk commented:
/ok-to-test sha=184aea7

akash-codemonk commented:
> @akash-codemonk
> 
> * [x]  Widget selection box doesn't goes off in 1st click on welcome tutorial page on 6th step.
>   i) Button icon is getting highlighted only for the 1st time when the widget tab open is it expected?
>   ii)The widget box lines are not aligned properly as well. Please refer to the attached video:- https://jiju8jbmwa.vmaker.com/record/OSaCRbF5Tvju6P7F

expected behaviour

> * [x]  Widget tool-tip and widget search box are overlapping with each other. Please refer to the attached SS:- ![image.png](https://camo.githubusercontent.com/4c67b8e4b9f50f225ac2236a854a9d38ac3c5a8e87cee8e655694c469c19e157/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3632626164666463366633636462333466656665633437652f30663264343564342d663432322d346134622d623462362d333337626132643532623831)

Fixed

> * [x]  The border of the canvas seems truncated. Please refer to the attached SS:- ![image.png](https://camo.githubusercontent.com/0e9059aa6c925c0816c5df34639245597625bc4f3fc515c8bd525328627cd8d4/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3632626164666463366633636462333466656665633437652f32653162363135312d326136392d343663652d623762632d626332333363303035623931)

fixed

> * [x]  Widget box fails to disappear from the canvas after the widget selection, it stays on the canvas until and unless clicked again. Is this expected behavior? Please refer to the video:- https://jiju8jbmwa.vmaker.com/record/8L6asZ6nXPh2qcSr

expected.

> * [ ]  Copy the widget from the top nav bar and paste it below the viewport. Now delete it and undo the same widget. Observe that the correct widget is not getting highlighted. Please refer to the attached video:- https://jiju8jbmwa.vmaker.com/record/jDR8SVVRGLxHlvRx

Unable to repro

> * [x]  Widget save tool-tip stays on the canvas for a longer period of time than expected. Please refer the attached video:- https://jiju8jbmwa.vmaker.com/record/rFvxxBnGj9T3MQQW

Doesn't seem like a bug. Happens on release as well.

> * [x]  Widget menu box opens after every drag and drop of the widget from the top 5 widget list. Please refer the video:- https://jiju8jbmwa.vmaker.com/record/PXzj9wfaw7UtL9Zf

Fixed

> * [x]  widget tool-tip is getting truncated on the **Page Settings** page. Please refer the SS:- ![image.png](https://camo.githubusercontent.com/0b1d9e5413bf5f071594d1dd4cdf9af57cd0629077b729ff076f7623b93e18b6/68747470733a2f2f696d616765732e7a656e68756275736572636f6e74656e742e636f6d2f3632626164666463366633636462333466656665633437652f36333862353038332d343965612d346238632d613330642d383830383331363165346237)

Fixed

> * [ ]  When clicked on the **+** icon on the settings page the widget box doesn't open is this expected? Please refer the video:- https://jiju8jbmwa.vmaker.com/record/tXCLwHicNnMjc3l2

Unable to repro

> * [x]  Page settings pane completely disappears and the edit dialog still persists on the canvas. Please refer the attached video:- https://jiju8jbmwa.vmaker.com/record/8kAiwRvBWyIsSDyj

Existing issue on release. Out of scope for this PR.

> * [ ]  collapse the entity explorer and observe behavior of the plus button; refresh the page and check behavior. Please refer the video:- https://www.loom.com/share/1ca37de3e228476b9a9e1ac55d87ae69

Didn't understand the issue. Incorrect recording?
@Richarex  
Updated dp with fixes https://appsmith-m17a897n8-get-appsmith.vercel.app/

Richarex commented:
Verified these bugs on the latest DP, and they are working fine:
1. Widget selection box doesn't goes off in 1st click on welcome tutorial page on 6th step - Fixed
2. Widget tool-tip and widget search box are overlapping with each other - Fixed
3. The border of the canvas seems truncated - Fixed
4. widget tool-tip is getting truncated on the Page Settings page - Fixed
5. Widget save tool-tip stays on the canvas for a longer period of time than expected - Not a bug
6. Widget box fails to disappear from the canvas after the widget selection, it stays on the canvas until and unless clicked again - Not a bug (expected behavior)
7.  Page settings pane completely disappears and the edit dialog still persists on the canvas - Out of scope (existing on release)
8. When clicked on the + icon on the settings page the widget box doesn't open is this expected - fixed
9. Copy the widget from the top nav bar and paste it below the viewport. Now delete it and undo the same widget. Observe that the correct widget is not getting highlighted - Fixed
10. Widget menu box opens after every drag and drop of the widget from the top 5 widget list - Fixed

akash-codemonk commented:
/ok-to-test sha=53551e2

akash-codemonk commented on app/client/cypress/integration/Regression_TestSuite/ClientSideTests/ActionExecution/ClearStore_spec.ts:
```diff
@@ -3,7 +3,6 @@ import * as _objects from "../../../../support/Objects/ObjectsCore"
 describe("clearStore Action test", () => {
   before(() => {
     _objects.ee.DragDropWidgetNVerify("buttonwidget", 100, 100);
-    _objects.ee.NavigateToSwitcher("explorer");
```
No more tab switchers in explorer.

akash-codemonk commented on app/client/cypress/integration/Regression_TestSuite/ClientSideTests/ExplorerTests/Entity_Explorer_Widgets_Copy_Delete_Undo_spec.js:
```diff
@@ -61,6 +62,7 @@ describe("Test Suite to validate copy/delete/undo functionalites", function() {
     );
     // eslint-disable-next-line cypress/no-unnecessary-waiting
     cy.wait(500);
+    _.agHelper.GetNClick(_.ee._bindingsClose);
```
Need to close the bindings popup. Since the tab switcher is removed from the explorer. The entities in explorer are higher up than before so the popup was overlapping UI elements where the subsequent cypress commands were to run causing it to fail with `...<target> is behind element .... ` errors.

akash-codemonk commented on app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Linting/EntityPropertiesLint_spec.ts:
```diff
@@ -10,13 +10,13 @@ const jsEditor = ObjectsRegistry.JSEditor,
     const invalidProperty = "unknownProperty";
     // create Api1
     apiPage.CreateAndFillApi("https://jsonplaceholder.typicode.com/");
+    agHelper.BlurFocusedElement();
```
Same issue with the evaluated value popup overlapping elements where previously it was below the tabs switcher in explorer. So blurring to hide the eval popup

akash-codemonk commented on app/client/cypress/support/Pages/CanvasHelper.ts:
```diff
@@ -0,0 +1,62 @@
+    _droppableArea: "#div-dragarena-0",
+  };
+
+  public OpenWidgetPane() {
```
Cypress method to open the widget pane. Since the pane now opens or closes after drag n drop depending on where we drop the widget, this method opens the pane if it is closed else if it is open we ignore.

akash-codemonk commented on app/client/src/globalStyles/portals.ts:
```diff
@@ -41,7 +40,7 @@ export const PortalStyles = createGlobalStyle`
     filter: drop-shadow(0px 1px 3px rgba(16, 24, 40, 0.1)) drop-shadow(0px 1px 2px rgba(16, 24, 40, 0.06));
     transition: all 1s;
-    z-index: 3;
+    z-index: ${Layers.guidedTourOverlay};
```
Since the widget pane uses popover it's z index is higher. Also during guided tour we add a overlay on top of this, updated the z index to a higher value for it to be visible.

akash-codemonk commented on app/client/src/pages/Editor/EditorHeader.tsx:
```diff
@@ -182,7 +182,8 @@ const BindingBanner = styled.div`
   height: 36px;
   left: 50%;
-  top: ${(props) => props.theme.smallHeaderHeight};
+  top: ${(props) =>
```
snipe banner top margin update as we have a top bar now above the canvas.

akash-codemonk commented on app/client/src/pages/Editor/Explorer/EntityExplorer.tsx:
```diff
@@ -90,13 +95,18 @@ function EntityExplorer({ isActive }: { isActive: boolean }) {
   const showWidgetsSidebar = useCallback(() => {
     history.push(builderURL({ pageId }));
+    if (!pinned && active) {
+      dispatch(setExplorerActiveAction(false));
```
Collapse the explorer(when unpinned) when user clicks on add widget cta from the explorer as the widget pane cta will be behind the explorer.

akash-codemonk commented on app/client/src/pages/Editor/Explorer/Widgets/WidgetGroup.tsx:
```diff
@@ -62,7 +64,9 @@ export const ExplorerWidgetGroup = memo((props: ExplorerWidgetGroupProps) => {
   return (
     <Entity
-      addButtonHelptext={createMessage(ADD_WIDGET_TOOLTIP)}
+      addButtonHelptext={
```
Hide the add widget cta tooltip in the explorer when the widget pane is open to prevent it from overlapping with the widget pane.

akash-codemonk commented on app/client/src/pages/Editor/Explorer/index.tsx:
```diff
@@ -1,98 +1,14 @@
-        className={`flex-shrink-0 px-3 mt-1 py-2 border-t border-b border-[${Colors.Gallery}]`}
-      >
-        <Switcher activeObj={switches[activeSwitchIndex]} switches={switches} />
-      </div>
```
Removing explorer tab switcher.

akash-codemonk commented on app/client/src/pages/Editor/GlobalHotKeys/GlobalHotKeys.tsx:
```diff
@@ -411,7 +411,7 @@ const mapDispatchToProps = (dispatch: any) => {
     redo: () => dispatch(redoAction()),
     setPreviewModeAction: (shouldSet: boolean) =>
-      dispatch(setPreviewModeAction(shouldSet)),
+      dispatch(setPreviewModeInitAction(shouldSet)),
```
Fixes a bug with toggling preview mode with keyboard shortcut calling the incorrect action.

akash-codemonk commented on app/client/src/pages/Editor/GuidedTour/Guide.tsx:
```diff
@@ -33,7 +33,6 @@ import {
 } from "@appsmith/constants/messages";
 
 const GuideWrapper = styled.div`
-  margin-bottom: ${(props) => props.theme.spaces[4]}px;
```
Guided tour content need not add margin below. As the top bar adds the necessary margin now.

akash-codemonk commented on app/client/src/pages/Editor/GuidedTour/utils.ts:
```diff
@@ -190,6 +190,15 @@ export function highlightSection(
     const coordinates = getCoordinates(element);
+
+    // If the element is not visible fade off the border
+    if (!document.body.contains(element)) {
```
Fade out animation in guided tour as the widget pane can close now(previously in the explorer it always stayed open)

akash-codemonk commented on app/client/src/pages/Editor/ToggleModeButton.tsx:
```diff
@@ -29,7 +29,6 @@ const StyledButton = styled(Button)<{ active: boolean }>`
   `}
   padding: 0 ${(props) => props.theme.spaces[2]}px;
   color: ${Colors.GREY_900};
-  height: ${(props) => props.theme.smallHeaderHeight};
```
Height of the buttons was exceeding the top bar causing it to overflow causing the UI issue where the header border wasn't visible.

akash-codemonk commented on app/client/src/pages/Editor/WidgetsEditor/WidgetPaneCTA.tsx:
```diff
@@ -0,0 +1,186 @@
+    const hbufferOffset = 200;
+    // The distance from the left of the viewport + buffer + widget pane width
+    // If the cursor is in this area we don't open the pane after drop
+    const hOffset = ctaPosition.left + hbufferOffset + WIDGET_PANE_WIDTH;
```
Horizontally added an additional offset of 200 to hide/show the pane simply for better UX

akash-codemonk commented on app/client/src/pages/Editor/WidgetsEditor/WidgetPaneCTA.tsx:
```diff
@@ -0,0 +1,186 @@
+
+  useEffect(() => {
+    if (!pinned && active && openWidgetPanel) {
+      dispatch(forceOpenWidgetPanel(false));
```
To close the widget pane when the explorer is unpinned and opened

akash-codemonk commented on app/client/src/sagas/PageSagas.tsx:
```diff
@@ -1174,6 +1176,7 @@ function* setPreviewModeInitSaga(action: ReduxAction<boolean>) {
       }),
     );
   } else {
+    if (!inPreviewMode) return;
```
Fixes a bug with preview where pressing escape button on canvas would move the user to home screen. Pressing esc key toggles off preview mode. Here we do a back navigation when toggling off, so now we check if we are in preview mode only then do a back navigation.

akash-codemonk commented on app/client/src/selectors/editorSelectors.tsx:
```diff
@@ -302,6 +305,26 @@ export const getWidgetCards = createSelector(
+          commonWidgetTypes.indexOf(a.type) - commonWidgetTypes.indexOf(b.type),
+      );
+  },
+);
```
Get the config for the top 5 widgets.

akash-codemonk commented on app/client/src/widgets/CheckboxGroupWidget/component/index.tsx:
```diff
@@ -4,7 +4,6 @@ import { Alignment } from "@blueprintjs/core";
 
 import { Classes } from "@blueprintjs/core";
 import { ComponentProps } from "widgets/BaseComponent";
-import { Colors } from "constants/Colors";
```
Unused import

akash-codemonk commented:
/ok-to-test sha=53551e2

akash-codemonk commented:
/ok-to-test sha=52d214f
